### ISSUE TITLE
small memory leak

### ISSUE THREAD

**Issue by bruceravel** [jacobwilliams/json-fortran#25]
While testing some of my own code for memory problems using valgrind, I came across a small memory leak in json-fortran.  It's truly small -- 1 byte.  But I'm trying to make everything neat and tidy, so....

The leak can be seen by compiling the example under the header "Writing a JSON file" in the json-fortran README.md file.  I am using gfortran 4.9.2 on Ubuntu 14.04.

Running valgrind as

```
  valgrind --track-origins=yes --leak-check=full --show-leak-kinds=all ./example2
```

traces the leak to the first call to `json_initialize` and then to 

https://github.com/jacobwilliams/json-fortran/blob/master/src/json_module.f90#L920

The leak can be "fixed" by replacing the call to `json_clear_exceptions()` at line 889 with

``` fortran
    exception_thrown = .false.
```

that is, the leak is fixed by not setting

``` fortran
    err_message = ''
```

in the first call to `json_initialize`. 

Making this change has no obvious, detrimental impact on the example from the README.md file or on the example program that gets build when json-fortran is built.  (The same is true for my code :smiley: )

Finally, gfortran's `-g` flag is handy.  Without it, I would never have tracked down the problem.  You might consider using it in `build.sh`.

**Comment by zbeekman** [comment]
Move to blocked by vendor bug, @jacobwilliams? (Also my preference would be to move the blocked by vendor bug over to the left side of the waffle.io board, maybe even left of `Backlog` to denote that it’s very stuck (waiting on vendor fixes etc.)

**Comment by zbeekman** [comment]
?

**Comment by jacobwilliams** [comment]
I wanted to test this myself....haven't had time yet.

**Comment by zbeekman** [comment]
ok, no worries. Sorry to inundate you with a flood of comments etc on gh

**Comment by szaghi** [comment]
Oh no, your suggestions are **always** interesting!

Valgrind is great, but I notice that when used with some *too modern* fortran feature (OOP, unlimited polymorphic, deferred length string, lhs auto-reallocation...) some (minor) issues arise, e.g. false positive memory leaks. Anyhow, its integration into travis-ci is very interesting. Please, let me know of any progresses.

See you soon.

**Comment by zbeekman** [comment]
I’ve been looking through the code and am more and more convinced that there are lurking memory leaks, and that automatic reference counting might be the way to go, although this will be substantial work to implement.

**Comment by jacobwilliams** [comment]
It's possible... I've not encountered any that are too obvious, but that doesn't mean they are not there. Also, I would distinguish between memory leaks internal to the module vs. memory leaks caused by using the routines incorrectly (due to the pointers).  The latter are certainly possible.  The former should be fixable with the code as is.  I would say we should get the code coverage up to around 100% for the unit tests, and see about getting valgrind hooked up, and/or some other profiling tools.  If we can identify any internal memory leaks, we can fix them. We also need to have good tests for runtime speed, so we can know how major changes will effect speed (#75).  Also: I need to make sure the "correct" usage of all the routine are clearly documented so user error can be minimized.  My recommendation would be to get all that in place before looking at the ARC idea.

**Comment by zbeekman** [comment]
Yeah I think it makes sense to hold off on ARC for the time being too.

**Comment by zbeekman** [comment]
We should try to run the unit tests through valgrind sometime soon, to see if we can find any issues. I just installed a recent release of valgrind via home-brew on my laptop which *may* finally behave as expected. Otherwise, I have access to some other linux machines which should have gfortran 4.9.2 installed on them. I wonder if the recent improvements (speed, unicode, etc.) have introduced or fixed any compiler caused memory leaks or memory leaks caused by json-fortran code. I probably won’t have time for at least another week, but it’s on my todo list.

**Comment by zmiimz** [comment]
In my test I see only "bytes in 1 blocks are still reachable in loss record" with valgrind 
That is not memory leak at all,  because  `err_message` and `real_fmt` are module variables and have by default `save` attribute.

If one wants to get rid of such valgrind messages, it is enough to add 
```fortran
if(allocated(err_message)) deallocate(err_message)
if(allocated(real_fmt)) deallocate(real_fmt)
```
to appropriate destroy_xx subroutines of json derived types

**Comment by zbeekman** [comment]
we should follow up on this... It may not even be relevant after the latest big refactor... I have learned that TAU can also check for memory leaks and I am curious if it can handle the modern fortran with it's PDT parsers... I may try this out if I find some time...
